### PR TITLE
Account-managed wallet flow improvements

### PR DESCRIPTION
**Pull request by howard-stearns**
- Fix a bug where in some cases that you don’t have enough money to buy, we would grey out the confirmation button, but not message the user to say why. In that message, add a link to the bank.
- Instead of telling people to setup or activate wallet (when opening marketplace or when attempting to buy), we now just silently setup wallet when first opening marketplace, leaving the user on the marketplace page.

### CODE REVIEW

**Comment by howard-stearns** [comment]
https://highfidelity.fogbugz.com/f/cases/18507/wallet-demo-punchlist-setup-distraction-was-Wallet-fix-Follow-wallet-setup-referrer-when-autogenerating-new-wallet

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/13418/HighFidelity-Beta-PR14102-75de305.apk

**Comment by howard-stearns** [comment]
**Test plan:**

_Basic_: 
1. Login with a new, never-before-used account. (Presumably on staging.)
2. Open the marketplace. There should be a short flash (of the wallet screen) only this first time, and then it should continue loading the marketplace page as normal.
3. All operations should be normal from then on, including all wallet market and wallet operations.

_Insufficient Funds_:
1. Try to buy one of each kind of content (avatar, entity, wearable, app, content set) without having enough money. You should get an explanatory text with a link to the bank. (The "bank" doesn't actually exist in staging.)
2. Get some money and then buy one of each. That should work normally.
3. Now repeat 1 for the same set of items (giving away your money if you have to, so that you don't have enough to buy again). It should tell you what's going on as in 1, but should give you the option of showing you the one you have in My Purchases.

_Regression_:
1. Everything else in the commerce tests should work as before.

File: scripts/system/marketplaces/marketplaces.js [hunk 1]
```diff
@@ -88,6 +88,46 @@ function setTabletVisibleInSecondaryCamera(visibleInSecondaryCam) {
 function openWallet() {
     ui.open(MARKETPLACE_WALLET_QML_PATH);
 }
+function setupWallet(cta) {
```
Comment (zfox23) [inline scripts/system/marketplaces/marketplaces.js]: Is the argument name `cta` valid here? Isn't it actually `referrer`?

File: scripts/system/marketplaces/marketplaces.js [hunk 2]
```diff
@@ -88,6 +88,46 @@ function setTabletVisibleInSecondaryCamera(visibleInSecondaryCam) {
...
+    // Needs to be done within the QML page in order to get access to QmlCommerce
+    openWallet();
+    var ALLOWANCE_FOR_EVENT_BRIDGE_SETUP = 0;
```
Comment (zfox23) [inline scripts/system/marketplaces/marketplaces.js]: The test plan should note that the tester should execute the plan on a min-spec machine (if available). We want to make sure `0` works in as many cases as possible.

File: scripts/system/marketplaces/marketplaces.js [hunk 3]
```diff
@@ -88,6 +88,46 @@ function setTabletVisibleInSecondaryCamera(visibleInSecondaryCam) {
...
+
+function openMarketplace(optionalItemOrUrl) {
+    // This is a bit of a kluge, but so is the whole file.
+    // If given a whole path, use it with no cta.
+    // If given an id, build the appropriate url and use the id as the cta.
+    // Otherwise, use home and 'marketplace cta'.
+    // AND... if call onMarketplaceOpen to setupWallet if we need to.
+    var url = optionalItemOrUrl || MARKETPLACE_URL_INITIAL;
+    var cta = 'marketplace cta';
+    if (optionalItemOrUrl && optionalItemOrUrl.indexOf(METAVERSE_SERVER_URL) === -1) {
```
Comment (zfox23) [inline scripts/system/marketplaces/marketplaces.js]: Why are we looking for `METAVERSE_SERVER_URL`? The comment below in the `else if` case makes me think you want the `else if` code to get executed when we're looking at a specific item, but won't the code go through this branch in that case, too?

File: scripts/system/marketplaces/marketplaces.js [hunk 4]
```diff
@@ -88,6 +88,46 @@ function setTabletVisibleInSecondaryCamera(visibleInSecondaryCam) {
...
+        cta = optionalItemOrUrl;  // item id
+        url = MARKETPLACE_URL + '/items/' + optionalItemOrUrl;
+    } else if (optionalItemOrUrl) {
+        // A specific url. Don't attempt to set up wallet because we don't have a way to come back.
+        // Does that ever happen without the wallet already setup?
```
Comment (zfox23) [inline scripts/system/marketplaces/marketplaces.js]: Unless I'm misunderstanding this comment, that can happen when the user clicks an item title in an inspection certificate.

File: scripts/system/marketplaces/marketplaces.js [hunk 5]
```diff
@@ -780,6 +821,14 @@ var onQmlMessageReceived = function onQmlMessageReceived(message) {
         return;
     }
     switch (message.method) {
+    case 'gotoBank':
+	ui.close();
```
Comment (zfox23) [inline scripts/system/marketplaces/marketplaces.js]: weird spacing

File: scripts/system/marketplaces/marketplaces.js (hunk 1 above)
Comment (howard-stearns) [reply scripts/system/marketplaces/marketplaces.js]: you're right. I'll change it.

File: scripts/system/marketplaces/marketplaces.js (hunk 2 above)
Comment (howard-stearns) [reply scripts/system/marketplaces/marketplaces.js]: good point.

File: scripts/system/marketplaces/marketplaces.js (hunk 5 above)
Comment (howard-stearns) [reply scripts/system/marketplaces/marketplaces.js]: wilco

File: scripts/system/marketplaces/marketplaces.js (hunk 3 above)
Comment (howard-stearns) [reply scripts/system/marketplaces/marketplaces.js]: I should add a comment. This branch is for item ids (and not whole urls even if they are already formatted to include items). If optionalItemOrUrl does not have our domain, then it isn't a url - just an item id. (Note `=== -1`, not `>= -1`.)

File: scripts/system/marketplaces/marketplaces.js [hunk 6]
```diff
@@ -626,6 +670,8 @@ var filterText; // Used for updating Purchases QML
 function onWebEventReceived(message) {
     message = JSON.parse(message);
     if (message.type === GOTO_DIRECTORY) {
+	// This is the chooser between marketplaces. Only OUR markteplace
+	// requires/makes-use-of wallet, so doesn't go through openMarketplace bottleneck.
```
Comment (howard-stearns) [inline scripts/system/marketplaces/marketplaces.js]: more whitespace issues

File: scripts/system/marketplaces/marketplaces.js (hunk 4 above)
Comment (howard-stearns) [reply scripts/system/marketplaces/marketplaces.js]: Good point. So what should we do there if the user is logged in and does not have their wallet?

We could try to set up their wallet, but we wouldn't know where to send them back to. I guess we could try to parse out of the url, but that seems fragile. Sigh...

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/13432/HighFidelity-Beta-PR14102-a6ef25f.apk

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/13432/HighFidelity-Beta-PR14102-a6ef25f.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/13432/HighFidelity-Beta-Interface-PR14102-a6ef25f.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/13432/HighFidelity-Beta-PR14102-a6ef25f.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/13432/HighFidelity-Beta-Interface-PR14102-a6ef25f.exe

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/13438/HighFidelity-Beta-PR14102-b45677c.apk

File: scripts/system/marketplaces/marketplaces.js (hunk 3 above)
Comment (zfox23) [reply scripts/system/marketplaces/marketplaces.js]: (I'm not going to resolve since you mentioned an action item, "add a comment".)

**Comment by rickcarr044** [comment]
Currently testing PR14102

**Comment by howard-stearns** [comment]
rats. I see a bug!  don't test yet!

**Comment by rickcarr044** [comment]
Stopped testing PR14102

**Comment by howard-stearns** [comment]
ok, I think I've fixed the bug

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/13462/HighFidelity-Beta-PR14102-053f221.apk

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/13462/HighFidelity-Beta-PR14102-053f221.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/13462/HighFidelity-Beta-Interface-PR14102-053f221.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/13462/HighFidelity-Beta-PR14102-053f221.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/13462/HighFidelity-Beta-Interface-PR14102-053f221.exe

**Comment by qlilaseguinolaza** [comment]
Currently testing

**Pull request merged**